File: Chapter 08/JS/section_8_2_1_declarative_querying/demo2.js
// Section2_DeclarativeQuerying/demo2.js
import { PrismaClient } from '@prisma/client';

// Initialize the Prisma Client (This replaces 'new AppDbContext()')
const prisma = new PrismaClient();

export class Demo2 {
    static async runQueryComparison() {
        console.log("\n=== Section 8.2.1: Declarative Querying (Raw SQL vs ORM) ===");
        console.log("SCENARIO: The database contains 4 users. We need to find all active users over age 21, sorted alphabetically.");

        // 1. Setup: Clean the database for the demo
        await prisma.user.deleteMany({});

        // 2. Seed the database with test data
        await prisma.user.createMany({
            data: [
                { firstName: "Alice", lastName: "Smith", age: 25, isActive: true },
                { firstName: "Bob", lastName: "Jones", age: 19, isActive: true },      // Fails: Too young
                { firstName: "Charlie", lastName: "Brown", age: 30, isActive: false }, // Fails: Inactive
                { firstName: "Diana", lastName: "Prince", age: 28, isActive: true }
            ]
        });
        console.log("SETUP: 4 Users inserted into the database.\n");

        // --- THE OLD WAY (IMPERATIVE) ---
        console.log("--- 1. The Old Way (Imperative / Raw SQL) ---");
        const rawSql = `SELECT * FROM users WHERE age > 21 AND is_active = 1 ORDER BY last_name ASC`;
        console.log(`  [Action] Executing Raw String: ${rawSql}`);
        
        // We force Prisma to execute a raw string
        const rawUsers = await prisma.$queryRawUnsafe(rawSql);
        
        const foundRaw = rawUsers.map(u => `${u.first_name} ${u.last_name}`).join(", ");
        console.log(`  [Result] Found: [${foundRaw}]`);
        console.log("  [Lesson] The burden is on you. If you mistyped 'is_active' as 'active' inside that string,");
        console.log("           your code would compile perfectly, but crash in production.\n");

        // --- THE MODERN WAY (DECLARATIVE) ---
        console.log("--- 2. The Modern Way (Declarative / ORM) ---");
        console.log("  [Action] Building a query object using Prisma's declarative syntax...");
        
        // This is the Prisma equivalent of Listing 8.4 from the textbook!
        const ormUsers = await prisma.user.findMany({
            where: {
                age: { gt: 21 },
                isActive: true
            },
            orderBy: {
                lastName: 'asc'
            }
        });

        const foundOrm = ormUsers.map(u => `${u.firstName} ${u.lastName}`).join(", ");
        console.log(`  [Result] Found: [${foundOrm}]`);
        console.log("  [Lesson] The ORM translates your JS object into SQL safely behind the scenes.");
        console.log("           If you rename 'age' to 'yearsOld' in your schema, this code");
        console.log("           will instantly throw a compilation error, saving you from a crash.");
    }
}

File: Chapter 08/C#/Section_8_2_1_Declarative_Querying/Demo2.cs
// Section2_DeclarativeQuerying/Demo2.cs
using System;
using System.Linq;
using Microsoft.EntityFrameworkCore;

namespace Chapter08.DeclarativeQuerying
{
    public class Demo2
    {
        public static void RunQueryComparison()
        {
            Console.WriteLine("\n=== Section 8.2.1: Declarative Querying (Raw SQL vs ORM) ===");
            Console.WriteLine("SCENARIO: The database contains 4 users. We need to find all active users over age 21, sorted alphabetically.");

            using var context = new AppDbContext();
            
            // 1. Setup: Reset the database and let the ORM build the tables automatically
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            // 2. Seed the database with test data
            context.Users.AddRange(
                new User { FirstName = "Alice", LastName = "Smith", Age = 25, IsActive = true },
                new User { FirstName = "Bob", LastName = "Jones", Age = 19, IsActive = true },      // Fails: Too young
                new User { FirstName = "Charlie", LastName = "Brown", Age = 30, IsActive = false }, // Fails: Inactive
                new User { FirstName = "Diana", LastName = "Prince", Age = 28, IsActive = true }
            );
            context.SaveChanges();
            Console.WriteLine("SETUP: 4 Users inserted into the database.\n");

            // --- THE OLD WAY (IMPERATIVE) ---
            Console.WriteLine("--- 1. The Old Way (Imperative / Raw SQL) ---");
            string rawSql = "SELECT * FROM Users WHERE Age > 21 AND IsActive = 1 ORDER BY LastName";
            Console.WriteLine($"  [Action] Executing Raw String: {rawSql}");
            
            // We force EF Core to execute a raw string
            var rawUsers = context.Users.FromSqlRaw(rawSql).ToList();
            
            var foundRaw = string.Join(", ", rawUsers.Select(u => $"{u.FirstName} {u.LastName}"));
            Console.WriteLine($"  [Result] Found: [{foundRaw}]");
            Console.WriteLine("  [Lesson] The burden is on you. If you mistyped 'IsActive' as 'Active' inside that string,");
            Console.WriteLine("           your code would compile perfectly, but crash in production.\n");

            // --- THE MODERN WAY (DECLARATIVE) ---
            Console.WriteLine("--- 2. The Modern Way (Declarative / ORM) ---");
            Console.WriteLine("  [Action] Building a query object using native C# LINQ syntax...");
            
            // This is Listing 8.4 from the textbook!
            var ormUsers = context.Users
                .Where(u => u.Age > 21 && u.IsActive)
                .OrderBy(u => u.LastName)
                .ToList();

            var foundOrm = string.Join(", ", ormUsers.Select(u => $"{u.FirstName} {u.LastName}"));
            Console.WriteLine($"  [Result] Found: [{foundOrm}]");
            Console.WriteLine("  [Lesson] The ORM translates your C# LINQ into SQL safely behind the scenes.");
            Console.WriteLine("           If you rename 'User.Age' to 'User.YearsOld', Visual Studio will instantly flag the error.");
        }
    }
}